File: .claude/skills/agentteams/scripts/py/src/agentteams/slicer.py
"""Token estimator + budget profiles + greedy topological packer + group renderer."""
from __future__ import annotations

from pathlib import Path
from typing import Any

BUDGET_PROFILES: dict[str, dict[str, Any]] = {
    "conservative": {"maxTokens": 120000, "recommendedFeaturesPerGroup": [1, 3]},
    "comfortable": {"maxTokens": 180000, "recommendedFeaturesPerGroup": [2, 4]},
    "aggressive": {"maxTokens": 250000, "recommendedFeaturesPerGroup": [3, 5]},
}

CHAR_TO_TOKEN = 0.27
PLANNING_OVERHEAD = 0.15
TOOL_CALL_OVERHEAD = 0.20


class BudgetExceededError(Exception):
    code = "BUDGET_EXCEEDED"


def _file_chars(p: str) -> int:
    try:
        return Path(p).stat().st_size
    except FileNotFoundError:
        return 0


def estimate_task(task: dict[str, Any]) -> int:
    base = _file_chars(task["path"])
    tokens = base * CHAR_TO_TOKEN
    tokens *= 1 + PLANNING_OVERHEAD
    tool_calls = len(task.get("targetFiles") or [])
    tokens *= 1 + TOOL_CALL_OVERHEAD * tool_calls
    return round(tokens)


def estimate_story(story: dict[str, Any]) -> int:
    own = round(_file_chars(story["path"]) * CHAR_TO_TOKEN * (1 + PLANNING_OVERHEAD))
    tasks = sum(estimate_task(t) for t in (story.get("tasks") or []))
    return own + tasks


def estimate_feature(feature: dict[str, Any]) -> int:
    own = round(_file_chars(feature["path"]) * CHAR_TO_TOKEN * (1 + PLANNING_OVERHEAD))
    stories = sum(estimate_story(s) for s in feature["stories"])
    return own + stories


def pack_groups(plan: dict[str, Any], profile_name: str) -> list[dict[str, Any]]:
    profile = BUDGET_PROFILES.get(profile_name)
    if not profile:
        raise ValueError(f"unknown-profile: {profile_name}")
    budget = profile["maxTokens"]
    tolerance = budget * 1.10
    groups: list[dict[str, Any]] = []
    current: dict[str, Any] = {"features": [], "tokens": 0}

    for feature in plan["features"]:
        f_tokens = estimate_feature(feature)
        if f_tokens > tolerance:
            if current["features"]:
                groups.append(current)
                current = {"features": [], "tokens": 0}
            sub_feature = {**feature, "stories": []}
            sub_tokens = 0
            for story in feature["stories"]:
                s_tokens = estimate_story(story)
                if s_tokens > tolerance:
                    raise BudgetExceededError(
                        f"budget-exceeded: story {feature['id']}.{story['id']} (~{s_tokens}t) > {budget}t profile"
                    )
                if sub_tokens + s_tokens > tolerance:
                    groups.append({"features": [{**sub_feature, "_splitContinuation": True}], "tokens": sub_tokens})
                    sub_feature = {**feature, "stories": []}
                    sub_tokens = 0
                sub_feature["stories"].append(story)
                sub_tokens += s_tokens
            if sub_feature["stories"]:
                groups.append({"features": [sub_feature], "tokens": sub_tokens})
            continue
        if current["tokens"] + f_tokens > tolerance:
            groups.append(current)
            current = {"features": [], "tokens": 0}
        current["features"].append(feature)
        current["tokens"] += f_tokens
    if current["features"]:
        groups.append(current)
    return [{**g, "n": i + 1} for i, g in enumerate(groups)]


def render_group(group: dict[str, Any], plan: dict[str, Any], plan_dir: str, profile_name: str) -> str:
    feature_summary = "\n".join(
        f"- **F{f['id']}** — {f['slug']} ({len(f['stories'])} stories)" for f in group["features"]
    )
    acceptance_lines = "\n".join(
        f"- [ ] {a}" for f in group["features"] for a in f["acceptanceCriteria"]
    ) or "- [ ] (no acceptance criteria declared in member features)"
    next_n = group["n"] + 1
    resume_prompt = f"implement group {next_n} of {plan_dir} per {plan_dir}/.ai-harness/session-plan.md"
    titles = " + ".join(f["slug"] for f in group["features"])
    return f"""# Group {group['n']} — {titles}

## Goal

Implement {len(group['features'])} feature(s) of the plan as a single context-budgeted unit.

## Features

{feature_summary}

## Acceptance

{acceptance_lines}

## Token Estimate

~{group['tokens']:,} tokens (profile: {profile_name})

## Resume Prompt

```
{resume_prompt}
```
"""

File: .claude/skills/agentteams/scripts/mjs/src/slicer.mjs
// Token estimator + budget profiles + greedy topological packer + group renderer.
import fs from 'node:fs';

export const BUDGET_PROFILES = Object.freeze({
  conservative: { maxTokens: 120000, recommendedFeaturesPerGroup: [1, 3] },
  comfortable: { maxTokens: 180000, recommendedFeaturesPerGroup: [2, 4] },
  aggressive: { maxTokens: 250000, recommendedFeaturesPerGroup: [3, 5] },
});

const CHAR_TO_TOKEN = 0.27;
const PLANNING_OVERHEAD = 0.15;
const TOOL_CALL_OVERHEAD = 0.20;

function fileChars(p) {
  try {
    return fs.statSync(p).size;
  } catch {
    return 0;
  }
}

export function estimateTask(task) {
  const baseChars = fileChars(task.path);
  let tokens = baseChars * CHAR_TO_TOKEN;
  tokens *= 1 + PLANNING_OVERHEAD;
  const toolCalls = (task.targetFiles || []).length;
  tokens *= 1 + TOOL_CALL_OVERHEAD * toolCalls;
  return Math.round(tokens);
}

export function estimateStory(story) {
  const own = Math.round(fileChars(story.path) * CHAR_TO_TOKEN * (1 + PLANNING_OVERHEAD));
  const tasks = (story.tasks || []).reduce((acc, t) => acc + estimateTask(t), 0);
  return own + tasks;
}

export function estimateFeature(feature) {
  const own = Math.round(fileChars(feature.path) * CHAR_TO_TOKEN * (1 + PLANNING_OVERHEAD));
  const stories = feature.stories.reduce((acc, s) => acc + estimateStory(s), 0);
  return own + stories;
}

export function packGroups(plan, profileName) {
  const profile = BUDGET_PROFILES[profileName];
  if (!profile) {
    throw Object.assign(new Error(`unknown-profile: ${profileName}`), { code: 'MISUSE' });
  }
  const budget = profile.maxTokens;
  const groups = [];
  let current = { features: [], tokens: 0 };
  const tolerance = budget * 1.10;

  for (const feature of plan.features) {
    const fTokens = estimateFeature(feature);
    if (fTokens > tolerance) {
      // Need to split — emit prior, then split-by-story
      if (current.features.length > 0) {
        groups.push(current);
        current = { features: [], tokens: 0 };
      }
      let subFeature = { ...feature, stories: [] };
      let subTokens = 0;
      for (const story of feature.stories) {
        const sTokens = estimateStory(story);
        if (sTokens > tolerance) {
          const err = new Error(`budget-exceeded: story ${feature.id}.${story.id} (~${sTokens}t) > ${budget}t profile`);
          err.code = 'BUDGET_EXCEEDED';
          throw err;
        }
        if (subTokens + sTokens > tolerance) {
          groups.push({ features: [{ ...subFeature, _splitContinuation: true }], tokens: subTokens });
          subFeature = { ...feature, stories: [] };
          subTokens = 0;
        }
        subFeature.stories.push(story);
        subTokens += sTokens;
      }
      if (subFeature.stories.length > 0) {
        groups.push({ features: [subFeature], tokens: subTokens });
      }
      continue;
    }
    if (current.tokens + fTokens > tolerance) {
      groups.push(current);
      current = { features: [], tokens: 0 };
    }
    current.features.push(feature);
    current.tokens += fTokens;
  }
  if (current.features.length > 0) groups.push(current);
  return groups.map((g, i) => ({ ...g, n: i + 1 }));
}

export function renderGroup(group, plan, planDir, profileName) {
  const slug = group.features[0]?.slug || `group-${group.n}`;
  const featureSummary = group.features.map(f => `- **F${f.id}** — ${f.slug} (${f.stories.length} stories)`).join('\n');
  const acceptanceLines = group.features.flatMap(f => f.acceptanceCriteria).map(a => `- [ ] ${a}`).join('\n');
  const nextGroup = group.n + 1;
  const resumePrompt = `implement group ${nextGroup} of ${planDir} per ${planDir}/.ai-harness/session-plan.md`;

  return `# Group ${group.n} — ${group.features.map(f => f.slug).join(' + ')}

## Goal

Implement ${group.features.length} feature(s) of the plan as a single context-budgeted unit.

## Features

${featureSummary}

## Acceptance

${acceptanceLines || '- [ ] (no acceptance criteria declared in member features)'}

## Token Estimate

~${group.tokens.toLocaleString()} tokens (profile: ${profileName})

## Resume Prompt

\`\`\`
${resumePrompt}
\`\`\`
`;
}
